lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs

```diff
@@ -563,33 +560,33 @@ balanceTransaction
             then balanceWith SelectionStrategyMinimal
             else throwE e
   where
+    era = recentEra @era
+
     -- Creates an index of all UTxOs that are already spent as inputs of the
     -- partial transaction.
     --
     -- This function will fail if any of the inputs refers to a UTxO that
     -- cannot be found in the UTxO reference set.
     --
-    extractExternallySelectedUTxO
+    indexPreselectedUTxO
         :: ExceptT (ErrBalanceTx era) m (UTxOIndex.UTxOIndex WalletUTxO)
-    extractExternallySelectedUTxO = do
-        let res = flip map txIns $ \i ->
-                case txinLookup i utxoReference of
-                    Nothing ->
-                       Left i
-                    Just o -> do
-                        let i' = Convert.toWallet i
-                        let W.TxOut addr bundle = toWalletTxOut era o
-                        pure (WalletUTxO i' addr, bundle)
-        case partitionEithers res of
-            ([], resolved) -> pure $ UTxOIndex.fromSequence resolved
-            (unresolvedInsHead : unresolvedInsTail, _) ->
-                throwE
-                . ErrBalanceTxUnresolvedInputs
-                $ (unresolvedInsHead :| unresolvedInsTail)
+    indexPreselectedUTxO
+        | Just unresolvedTxIns <- maybeUnresolvedTxIns =
+            throwE (ErrBalanceTxUnresolvedInputs unresolvedTxIns)
+        | otherwise = pure $
+            UTxOIndex.fromSequence (convertUTxO <$> Map.toList selectedUTxO)
       where
-        era = recentEra @era
-        txIns :: [TxIn]
-        txIns = Set.toList $ tx ^. bodyTxL . inputsTxBodyL
+        convertUTxO :: (TxIn, TxOut era) -> (WalletUTxO, W.TokenBundle)
+        convertUTxO (i, o) = (WalletUTxO (Convert.toWallet i) addr, bundle)
+          where
+            W.TxOut addr bundle = toWalletTxOut era o
+        maybeUnresolvedTxIns :: Maybe (NonEmpty TxIn)
+        maybeUnresolvedTxIns =
+            NE.nonEmpty $ Set.toList $ txIns <\> Map.keysSet selectedUTxO
+        selectedUTxO :: Map TxIn (TxOut era)
+        selectedUTxO = Map.restrictKeys (unUTxO utxoReference) txIns
```

Comment:
`Map.restrictKeys` is presumably decently efficient with large utxoReference and small txIns?